ciri: exit on CTRL-C

howjmay opened the pull request:
call core_stop(&core_g) and core_destroy(&core_g) if pressed ctrl-C, close #133

howjmay committed:
implement keyboard interrupt

thibault-martinez commented:
Hi @HowJMay ,

- We probably need to abstract signals as their API might differ from one platform to another (take a look at `/utils/handles/*`)
- What you are currently doing is checking a flag _after_ the sleep. This sleep is temporary and replace the missing running API that will _never_ finish so it cannot work like that.

howjmay commented:
Hi @thibault-martinez I push a new version code to exit cIRI.
In this version, once user press ctrl-C, the signal handler will call `core_stop(&core_g)` and `core_destroy(&core_g)`
Is this version a better implementation? Or I need to implementation the exit in the other way?
BTW, in this version, I haven't saloved the abstract signals problem yet.
Thank you for your kindness.

thibault-martinez commented:
- Registering your signal is probably something you want to do as soon as possible in the main
- No need to destroy in the handler, `stop` will trigger `start` and destroy will follow on the main

jkrvivian commented on ciri/main.c:
```diff
@@ -17,6 +19,21 @@
+  if (signo == SIGINT) {
+	  log_info(MAIN_LOGGER_ID, "Stopping cIRI core\n");
+    if (core_stop(&core_g)) {
+      log_error(MAIN_LOGGER_ID, "Stopping cIRI core failed\n");
```
Did you run the git commit hook?

HowJMay replied:
Yes I did.
Is there anything wrong?

howjmay replied:
I will fix it sorry

howjmay committed:
implement abstract signal handler

thibault-martinez commented on utils/handles/sig.c:
```diff
@@ -0,0 +1,22 @@
+
+void signal_handler_posix(int signo) {
+  if (signo == ctrl_c_signal) {
+    signal_handler_core_end(signo);
```
This API shouldn't be linked in any way to cIRI functions, it should be usable in any case.

thibault-martinez commented on utils/handles/sig.h:
```diff
@@ -0,0 +1,65 @@
+/*
```
Call this file `signal.h` please

thibault-martinez commented on utils/handles/sig.h:
```diff
@@ -0,0 +1,65 @@
+ * We declare a function register_signal() to a system available condition
+ * variable primitive and its associated functions, some of them might have no
+ * effect if not needed by the underlying API
+ */
```
Be careful when copy-pasting

howjmay replied:
sorry for this
I will fix it asap

thibault-martinez commented on utils/handles/sig.h:
```diff
@@ -0,0 +1,65 @@
+
+#else
+
+#error "No lock primitive found"
```
Be careful when copy-pasting

howjmay committed:
delink connection between main.c and signal.h

Make the APIs in signal.h is independent of main.c.

Rename sig.h and sig.c into signal.h, signal.c respectively

howjmay committed:
Correct copy-pasting mistakes

thibault-martinez commented on ciri/main.c:
```diff
@@ -5,18 +5,29 @@
  * Refer to the LICENSE file for licensing information
  */
 
+#include <signal.h>
```
I think this is not needed here anymore

thibault-martinez commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,71 @@
+
+static inline __sighandler_t register_signal(
+    int SIG, void (*handler_core_end)(universal_singal_num_t)) {
+  end_core = handler_core_end;
```
Names shouldn't be related to core, they should be generic

howjmay replied:
Hi since I thought the std <signal.h> passes a function as signal handler, therefore, I imitated the std <signal.h> pass a function.
should I mvoe the signal handler in `main.c` to `utils/handles/signal.c` ?

thibault-martinez commented on utils/handles/signal.c:
```diff
@@ -0,0 +1,22 @@
+#include "signal.h"
```
Copyright

thibault-martinez commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,71 @@
+ */
+
+#ifndef __UTILS_HANDLES_SIG_H__
+#define __UTILS_HANDLES_SIG_H__
```
`__UTILS_HANDLES_SIGNAL_H__`

howjmay committed:
Correct define name and add copyright

thibault-martinez commented:
Let's get back to this, sorry for the long wait.
Please have a look at #472 as it will influence this PR.
The main function will basically be printing logs at a given interval, so we will just need to set a boolean to false in the ctrl-c handle to stop it and proceed to core stop/destroy.
I'll add some comments on the code.

thibault-martinez commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,71 @@
+#include <signal.h>
+#include <stdlib.h>
+
+typedef enum UNIVERSAL_SINGAL_NUM { null, ctrl_c } universal_singal_num_t;
```
Typos

thibault-martinez commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,71 @@
+    (defined(__APPLE__) && defined(__MACH__))
+
+#include <unistd.h>
+#define ctrl_c_signal SIGINT
```
Uppercase defines

thibault-martinez commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,71 @@
+
+static inline __sighandler_t register_signal(
+    int SIG, void (*handler_core_end)(universal_singal_num_t)) {
+  end_core = handler_core_end;
```
This won't work if we want to register multiple signals

thibault-martinez replied:
Hi @HowJMay, did you have time to think about this ?
Don't worry if you're not available these days, just tell me and I'll take the issue.
Also, happy new year !

howjmay replied:
sorry I was not available in the past a few months, but now I am available now!
I will push a new commit in the recent few days, and Happy New Year haha =)

howjmay replied:
Sorry I get one problem. Do we need several different signal handlers which handles one signal respectively? Or I can use one signal handler to handle all signals?

thibault-martinez replied:
Oh, never saw this comment sorry. Do you still have a problem ? The UNIX way to do it is that a signal is bound to a handler but many signals can be bound to the same handler.

howjmay replied:
thank you for answering. I will do it now.

thibault-martinez replied:
We'd like this feature to be part of the current sprint so I'll take the time to review it and give you some guidance if needed. Feel free to discuss this in #ciri-discussion on Discord :)

howjmay replied:
Thank you for your kindness. 
I just implemented the signal interrupt for UNIX, but I met one problem.
I can build this new code successfully with `bazel build //ciri:ciri`
however, the execution file `ciri` can't run successfully. Did I use the wrong way to build ciri?

thibault-martinez replied:
I'm going to need a little bit more details. How do you run ciri and what's the error ? You can build and run ciri at the same time with the command `bazel run -- ciri`

howjmay replied:
the command help me fixing the problem, thank you!

thibault-martinez commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,71 @@
+void (*end_core)(universal_singal_num_t);
+
+static inline __sighandler_t register_signal(
+    int SIG, void (*handler_core_end)(universal_singal_num_t)) {
```
Call it `signal_handler`, not `handler_core_end `

thibault-martinez commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,71 @@
+void (*end_core)(universal_singal_num_t);
+
+static inline __sighandler_t register_signal(
+    int SIG, void (*handler_core_end)(universal_singal_num_t)) {
```
Call it `signal_handler`

thibault-martinez commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,71 @@
+         */
+
+static inline __sighandler_t register_signal(
+    int SIG, void (*handler_core_end)(universal_singal_num_t));
```
Call it `signal_handler`

thibault-martinez commented on ciri/main.c:
```diff
@@ -11,12 +11,22 @@
 
 static core_t core_g;
 
+void signal_handler_core_end(universal_singal_num_t signo) {
```
Just set a boolean to false here

howjmay commented:
hi sorry for the errors. I will do my best to fix them

howjmay committed:
fix typo and few erros

howjmay committed:
add brackets

howjmay committed:
Merge branch 'develop' of github.com:iotaledger/entangled into develop

howjmay committed:
Implement UNIX multiple signal interrupt

developer can register the signal to catch with
`register_signal(universal_signal_num_t SIG, bool (*register_signal_handler)())`

no matter which platform is the signal number can use universal_signal_num_t
consistently.

howjmay committed:
implement exit with ctrl-c on UNIX

Current code can terminate ciri after "starting API".
However, if API has not started yet, pressing ctrl-c will cause failure,
which is the ciri can't be terminated.

and windows version hasn't finished implementing yet.

jserv commented on utils/handles/signal.c:
```diff
@@ -13,31 +13,30 @@
+  switch (fdwCtrlType) {
+    // Handle the CTRL-C signal.
+    case CTRL_C_EVENT:
+      printf("Ctrl-C event\n\n");
```
Don't invoke `printf`. Use logging functions for diagnosis.

jserv commented on utils/handles/signal.c:
```diff
@@ -0,0 +1,42 @@
+ * handle.
+ * @para void (*signal_handler)(): signal handler of each signal needs
+ */
+__sighandler_t register_signal(universal_signal_num_t SIG,
```
Since `register_signal` is just a wrapper of OS-specific functions, you can move it to header `utils/handles/signal.h` in the form of `static inline` functions which can be optimized by compilers.

jserv commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,70 @@
+
+#else
+
+#error "No signal process library found"
```
Ditto. Implementing this in header file can validate supported platforms in advance.

howjmay replied:
So I should change this into logging functions?  Or I should add logging function and preserve the #error directive?

jserv replied:
Keeping `#error` for conditional build is reasonable since you already consider major UNIX-compatible systems and Windows-based environments.

jserv commented on utils/handles/signal.h:
```diff
@@ -0,0 +1,70 @@
+#include "utils/logger_helper.h"
+
+/*
+ *  This enum is declared for consistent name of signal for c5 ross platfrom
```
Typo. Should be "cross platform" instead of "c5 ross platform".

howjmay committed:
Merge branch 'develop' of github.com:iotaledger/entangled into develop

howjmay committed:
Merge branch 'develop' into ctrlC

jserv commented on ciri/main.c:
```diff
@@ -18,11 +19,45 @@
 static core_t core_g;
 static logger_id_t logger_id;
 
+void signal_handler() {
```
Should be marked as `static`.

jserv commented on utils/handles/signal.c:
```diff
@@ -0,0 +1,46 @@
+  }
+}
+
+static inline __sighandler_t register_signal(
```
static inline functions such as register_signal should appear in headers rather than C source.

jserv commented on utils/merkle.c:
```diff
@@ -19,10 +19,11 @@ static size_t binary_tree_size(size_t const acc, size_t const depth) {
 size_t merkle_size(size_t const leaf_count) {
   size_t acc = 1, count = leaf_count;
 
-  if (count == 0)
```
Is the change necessary?

jserv commented on utils/handles/signal.c:
```diff
@@ -0,0 +1,46 @@
+ * @para void (*signal_handler)(): signal handler of each signal needs
+ */
+static inline __sighandler_t register_signal(
+    universal_signal_num_t SIG, void (*register_signal_handler)()) {
```
Use lowercase `sig`.

howjmay committed:
fix errors

thibault-martinez committed:
Merge branch 'develop' into ctrlC

thibault-martinez committed:
ciri: reorganize sigint handling

thibault-martinez commented:
@HowJMay I pushed some changes to the PR.
Regarding the signal abstraction, I'm not familiar with Windows development but AFAIK `signal.h` and `SIGINT` are standard C so maybe we can just use them. I'll investigate tomorrow.

howjmay commented:
@thibault-martinez  thank you for correcting. I am not familiar with windows development either. however, the comment you gave helps me found some helpful information. thank you so much

howjmay committed:
implement windows signal catching

howjmay committed:
Merge branch 'ctrlC' of github.com:HowJMay/entangled into ctrlC

thibault-martinez committed:
utils: clean signal BUILD

thibault-martinez committed:
utils: fix signal_handle for macOS

thibault-martinez committed:
Merge branch 'develop' into ctrlC

thibault-martinez committed:
utils: fix non GNU code
